impala/hiveserver2.py

```diff
@@ -686,43 +690,56 @@ def fetch_results(service, operation_handle, hs2_protocol_version, schema=None,
     err_if_rpc_not_ok(resp)
 
     if hs2_protocol_version == TProtocolVersion.HIVE_CLI_SERVICE_PROTOCOL_V6:
+        from bitarray import bitarray
+
         tcols = [_TTypeId_to_TColumnValue_getters[schema[i][1]](col)
                  for (i, col) in enumerate(resp.results.columns)]
         num_cols = len(tcols)
         num_rows = len(tcols[0].values)
         log.debug('fetch_results: COLUMNAR num_cols=%s num_rows=%s tcols=%s',
                   num_cols, num_rows, tcols)
-        rows = []
-        for i in range(num_rows):
-            row = []
-            for j in range(num_cols):
-                type_ = schema[j][1]
-                values = tcols[j].values
-                nulls = tcols[j].nulls
-                if six.PY3 and isinstance(nulls, str):
-                    # thriftpy sometimes returns unicode instead of bytes
-                    nulls = nulls.encode('utf-8')
-                # i // 8 is the byte, i % 8 is position in the byte; get the
-                # int repr and pull out the bit at the corresponding pos
-                is_null = False
+
+        column_data = []
+        for j in range(num_cols):
+            type_ = schema[j][1]
+            nulls = tcols[j].nulls
+            values = tcols[j].values
+
+            if six.PY3 and isinstance(nulls, str):
+                # thriftpy sometimes returns unicode instead of bytes
+                nulls = nulls.encode('utf-8')
+
+            # HACK: Hive hack; needs tests
+            # Hive encodes nulls differently than Impala
+            # (\x00 vs \x00\x00 ...)
+            if not re.match(b'^(\x00)+$', nulls):
```

Comment:
I think we can drop the regular expression as long as the nulls string is long enough